### PR TITLE
fix: detach should be sent for node

### LINKED ISSUE #17653: Detached node is not removed from the state tree when removeFromTree is called
### Description of the bug

When any element is removed via calling, `Element::removeFromTree`, the detached node still persists in the client-side tree. Therefore, it is never garbage collected, leading to potential memory leaks.

Follow-up from a Flow component [issue](https://github.com/vaadin/flow-components/issues/5427).

### Expected behavior

I would expect the detached node to be removed from the client side-tree.

### Minimal reproducible example

```java
Element element = new Element("vaadin-random-element");
UI.getCurrent().getElement().appendChild(element);
add(new Button("Remove all elements from tree", click -> element.removeFromTree()));
```
1. Add the snippet to a view
2. Click the button
3. Observe that the detached elements are still in the tree
4. Elements can be appended and removed repeatedly to reproduce a growing number of detached elements 

### Versions

- Vaadin / Flow version: 24.2
- Java version: 17
- OS version: MacOS
- Browser version (if applicable): Chrome 116
- Application Server (if applicable): -
- IDE (if applicable): -

### PR DESCRIPTION
**Pull request by caalador**
Do not clean the node id or
hasAttached state too early as
then we do not send the detach
event for the node.

Fixes #17653

### CODE REVIEW

File: flow-server/src/main/java/com/vaadin/flow/internal/change/NodeDetachChange.java [hunk 1]
```diff
@@ -18,6 +18,7 @@
 
 import com.vaadin.flow.internal.ConstantPool;
 import com.vaadin.flow.internal.StateNode;
+import com.vaadin.flow.server.Command;
```
Comment (mcollovati) [inline flow-server/src/main/java/com/vaadin/flow/internal/change/NodeDetachChange.java:21]: unused import
```suggestion
```

**Review by mcollovati — CHANGES_REQUESTED** [review]
Tested locally and could not see detached elements on the browser anymore after removing elements with `removeFromTree()`.

**Comment by mcollovati** [comment]
Looks like there are issues with PreserveOnRefresh

**Review by mcollovati — CHANGES_REQUESTED** [review]
Just some minor javadocs changes.
I wonder if we can somehow arrange unit tests for StateNode, to prevent regressions.
Maybe just checking for `id` and `wasAttached` (through `isClientSideInitialized`?) values after invoking `removeFromTree` with true and false.

File: flow-server/src/main/java/com/vaadin/flow/internal/StateNode.java [hunk 2]
```diff
@@ -387,24 +387,38 @@ protected void setTree(StateTree tree) {
...
+        visitNodeTree(node -> node.reset(sendDetach));
         setParent(null);
     }
 
     /**
      * Resets the node to the initial state where it is not owned by a state
      * tree.
      */
-    private void reset() {
+    private void reset(boolean shouldDetach) {
```
[the remark this answers is not in the document]
Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/internal/StateNode.java:413]: added

File: flow-server/src/main/java/com/vaadin/flow/dom/Element.java [hunk 3]
```diff
@@ -525,7 +525,10 @@ public Element removeFromParent() {
      * @return this element
      */
     public Element removeFromTree() {
+        return removeFromTree(true);
+    }
 
+    public Element removeFromTree(boolean sendDetach) {
```
[the remark this answers is not in the document]
Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/dom/Element.java:531]: added

**Review by mcollovati — APPROVED** [review]

**Pull request merged**